### ISSUE TITLE
SWC or Rslint ?

### ISSUE THREAD

**Issue by HerringtonDarkholme** [HerringtonDarkholme/vue-template-compiler#17]
TLDR: Rslitn is still better than swc.

The code changes are in swc branch and rslint branch.

swc contains a lot of dependencies regardless if they are relevant to core parsing. Such bloated dependencies place huge burden to compiling, slowing rust-analyzer to almost freezing.
swc's docs and examples re scarce. The only working example is https://rustdoc.swc.rs/swc_ecma_parser/.
Looking to the code is also hard. The abstraction and module organization is, well, hazy at least to the uneducated. Peeking the definition is hard, if possible, given the massive usage of macros. Alas, the macro is also the perpetrator of the sluggish compilation. :/ 
Using swc is not a nice journey, actually. Looking at the example above, it immediately requires several crates other than the parse. `common`, `ast`, `atom`, `visit` and blahblah. And the core `impl Visitor` has 200+ macro generated methods to implement without one single line of documentation. The usability is poor...

Rslint at least has more comments and documentation than swc. It's underlying crate, rowan, also has docs. So it might be a better choice.

**Comment by zhmushan** [comment]
How about [deno_lint](https://github.com/denoland/deno_lint)?

> A Rust crate for writing fast JavaScript and TypeScript linters.
>
> This crate powers [`deno lint`](https://deno.land/manual/tools/linter), but is
not Deno specific and can be used to write linters for Node as well.
>
> _Supports `recommended` set of rules from ESLint and `@typescript-eslint` out of
the box with no config._

**Comment by HerringtonDarkholme** [comment]
![image](https://user-images.githubusercontent.com/2883231/135373739-20c176b9-f3d4-4152-9119-350b88239803.png)

**Comment by HerringtonDarkholme** [comment]
I'm thinking about a more aggressive approach like gogocode or more advanced semgrep.
Basically I will take source code and feed it to tree-sitter.
Tree-sitter will give me an untyped syntax tree against which I can match a target syntax, via the type-id in the syntax node.

This approach is quite versatile and generic. But it might not be as fast as specialized syntax crates like rslint/swc.

**Comment by HerringtonDarkholme** [comment]
Reference:
https://github.com/brendanzab/moniker
https://github.com/flatt-security/shisho
https://github.com/thx/gogocode
https://semgrep.dev/

**Comment by phoenix-ru** [comment]
@HerringtonDarkholme 

I tried using both RSLint and SWC for the analysis and transformation parts. While doing analysis is somewhat easier with RSLint, I found its `Indel` API very quirky to use.

As for SWC, I actually implemented the analysis using it, and it is surprisingly a breeze. Just make sure you do not use `Visitor` APIs, as they make things way too complicated.

I mainly use the `swc_core` crate, which makes the binary size really manageable. The numbers are (for `--release`):
- 3.2 MB for the analyser crate which uses `swc_ecma_codegen`, `swc_ecma_parser`, `swc_core` and `lazy_static`. You can check the [Cargo.toml](https://github.com/phoenix-ru/rust-vue-compiler/blob/master/crates/fervid_script/Cargo.toml);
- (offtopic) 8.8 MB for the full compiler with `lightningcss` included ([Cargo.toml](https://github.com/phoenix-ru/rust-vue-compiler/blob/master/crates/fervid/Cargo.toml)).

As per performance, I did some micro-benchmarks using Criterion and found SWC to reliably perform 30%-50% better than RSLint. And while SWC was doing some transformations, RSLint was only parsing and not applying any transforms.

Here you can find the test suite of how analysis performs: 
https://github.com/phoenix-ru/rust-vue-compiler/blob/1d02ab364e222358ec9fdd713daf9afd7f92848d/crates/fervid_script/src/script_legacy/mod.rs#L110-L736

**Comment by HerringtonDarkholme** [comment]
But compiling swc is really slow. It slows down my development iteration speed much. 
Back then Rust Analyzer cannot support swc's macro as well.

Further, I personally don't like deeply nested pattern matching and dealing with SWC related JsWord stuff.
See
https://github.com/phoenix-ru/rust-vue-compiler/blob/1d02ab364e222358ec9fdd713daf9afd7f92848d/crates/fervid_script/src/script_legacy/utils.rs#L12-L57
Compared to 
https://github.com/HerringtonDarkholme/vue-compiler/blob/main/crates/sfc/src/script/vanilla_script.rs#L125-L158
